### PR TITLE
John's name in the copyright header.

### PR DESCRIPTION
**Pull request by dkorolev**

### CODE REVIEW

**Comment by dkorolev** [comment]
OOPS, I've just pushed to ```master```. So be it.

@deathbaba , @sompylasar : Beyond a one-line change with John's name, this pull request now contains the documentation for ```JSON()``` and ```JSONParse``` methods, as well as the scripts to generate it.

I thought of a way to have documentation auto-generated from the code that is:
* compiled, run and passes the tests,
* looks clean on GitHub Wiki (.md), and
* impossible to have diverge from the actual Bricks code.

I think the code here does the job. Alex & John, please take a look.

CC @kt97679: Kirill, may I ask you to take a quick look at ```gen-docu.sh``` script?

Thanks,
Dima

**Comment by sompylasar** [comment]
@dkorolev I'd add a link to the cerealize library's github.
Other ideas and code look good (except leaking MapsYouEventBase and friends).

**Comment by deathbaba** [comment]
LGTM

File: KnowSheet/scripts/gen-docu.sh [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+#    and by the top-level `make` that runs all tests at once and generates the coverage report.
+#
+# 3) Use the two spaces wisely.
+#    If something should be in the docu, it should be indented.
+#    If it should not be, it should not be indented.
+#    The `docu_*.*` files are excluded from `make indent`.
+
+set -u -e
+
+for i in $(for i in $(find . -type f -iname "docu_*.*" | grep -v ".noshit"); do
```
Comment (kt97679) [inline KnowSheet/scripts/gen-docu.sh]: Same variable used in internal and external for loops. Probably works here, but overall confusing and can result in errors if you would change this code.

File: KnowSheet/scripts/gen-docu.sh [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+             echo -e $(basename $i)"\t"$i;
```
Comment (kt97679) [inline KnowSheet/scripts/gen-docu.sh]: Not sure why you print 2 fields if you cut just 2nd field on next step.

File: KnowSheet/scripts/gen-docu.sh [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+           done | sort | cut -f2) ; do
+  echo $i >/dev/stderr
+  case ${i##*.} in
+  cc)
+    echo '```c++'
+    (grep '^  ' $i | sed "s/^  //g") || echo "// No documentation data in '$i'."
```
Comment (kt97679) [inline KnowSheet/scripts/gen-docu.sh]: IMHO you don't need () here unless I miss something.

File: KnowSheet/scripts/gen-docu.sh [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+#    If it should not be, it should not be indented.
+#    The `docu_*.*` files are excluded from `make indent`.
+
+set -u -e
+
+for i in $(for i in $(find . -type f -iname "docu_*.*" | grep -v ".noshit"); do
+             echo -e $(basename $i)"\t"$i;
+           done | sort | cut -f2) ; do
+  echo $i >/dev/stderr
+  case ${i##*.} in
```
Comment (kt97679) [inline KnowSheet/scripts/gen-docu.sh]: IMHO
case $i in 
*.cc)
is a little bit more readable.

**Comment by kt97679** [comment]
Added some comments.

File: KnowSheet/scripts/gen-docu.sh (hunk 1 above)
Comment (dkorolev) [inline KnowSheet/scripts/gen-docu.sh]: Fixed, thanks!

File: KnowSheet/scripts/gen-docu.sh (hunk 4 above)
Comment (dkorolev) [inline KnowSheet/scripts/gen-docu.sh]: Thanks, fixed!

File: KnowSheet/scripts/gen-docu.sh (hunk 3 above)
Comment (dkorolev) [inline KnowSheet/scripts/gen-docu.sh]: You're probably right. I'll keep those for readability though, same reason I keep parentheses in ```c++
if ((a & b) || c) {
  ...
}```.

**Comment by dkorolev** [comment]
@kt97679 : Thank you!

@sompylasar : Thank you! Wow, how could I miss the URL for Cereal? In my defense, this PR is to demonstrate the approach, not the demo.

@sompylasar , @deathbaba : OK to merge from my side. (I'll probably click "Merge" myself in a coupe hours upon starting to work on the next thing, but am showing respect to our tradition to not merge own code.)

Thanks,
Dima

**Comment by dkorolev** [comment]
The leak is actually intentional :-)

Thanks,
Dima

http://dimakorolev.com/ | http://www.linkedin.com/in/dimakorolev |
dmitry.korolev@gmail.com | +1 (312) 593 2783

**Pull request merged**